node_api/src/logic/validator.ts

```diff
@@ -209,4 +210,35 @@ export default class Validator {
     }
     return { isValid: valid, message: message };
   }
+
+  private isValidRoleGroups(data: any): { isValid: boolean; message?: string } {
+    let valid = false;
+    let message: string | undefined = "1 or more role groups are invalid.";
+    if (isObjectHasKeys(data, [IM.ROLE_GROUP])) {
+      for (let group in data[IM.ROLE_GROUP]) {
+        if (isObjectHasKeys(data[IM.ROLE_GROUP][group], [IM.GROUP_NUMBER])) {
+          if (Object.keys(data[IM.ROLE_GROUP][group]).length > 1) {
+            for (let roles in data[IM.ROLE_GROUP][group]) {
+              if (null !== data[IM.ROLE_GROUP][group][roles]["@id"] && "" !== data[IM.ROLE_GROUP][group][roles].name) {
+                valid = true;
+                message = undefined;
+              } else {
+                valid = false;
+                message = "1 or more role groups are invalid.";
+                break;
+              }
+            }
+          } else {
+            valid = false;
+            message = "1 or more role groups are invalid.";
+            break;
+          }
+        }
+      }
+    } else {
+      valid = true;
+      message = undefined;
+    }
+    return { isValid: valid, message: message };
+  }
```

Comment:
For this function, it may be cleaner to invert the top level "if" and exit early, thereby reducing the number of nesting levels (cyclomatic complexity).

Inside the for-loop, rather than setting variables and exiting, it again may be cleaner to exit straight out. This also means the `valid = true` logic becomes unneccesary, with a simple return at the end if nothing fails.  The result looks something like this (though I havent thoroughly checked the logic but I'm sure there's unit tests that would confirm if the logic is the same or not ;) )

```
  private isValidRoleGroups(data: any): { isValid: boolean; message?: string } {
    if (!isObjectHasKeys(data, [IM.ROLE_GROUP])) return { isValid: true };

    for (let group in data[IM.ROLE_GROUP]) {
      if (isObjectHasKeys(data[IM.ROLE_GROUP][group], [IM.GROUP_NUMBER])) {
        if (Object.keys(data[IM.ROLE_GROUP][group]).length <= 1) {
          return { isValid: false, message: "1 or more role groups are invalid." };
        } else {
          for (let roles in data[IM.ROLE_GROUP][group]) {
            if (null === data[IM.ROLE_GROUP][group][roles]["@id"] || "" === data[IM.ROLE_GROUP][group][roles].name) {
              return { isValid: false, message: "1 or more role groups are invalid." };
            }
          }
        }
      }
    }

    return { isValid: true };
  }
```